radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/NextCandidateForkPostProcessor.java

```diff
@@ -0,0 +1,104 @@
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.statecomputer;
+
+import com.radixdlt.constraintmachine.REProcessedTxn;
+import com.radixdlt.engine.PostProcessor;
+import com.radixdlt.engine.PostProcessorException;
+import com.radixdlt.statecomputer.forks.CandidateForkConfig;
+import com.radixdlt.statecomputer.forks.Forks;
+import com.radixdlt.statecomputer.forks.ForksEpochStore;
+import com.radixdlt.store.EngineStore;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * Checks whether the engine should switch to the next candidate fork. If so, adds nextForkName to
+ * the result metadata.
+ */
+public final class NextCandidateForkPostProcessor implements PostProcessor<LedgerAndBFTProof> {
+  private final CandidateForkConfig nextFork;
+  private final Supplier<ForksEpochStore> forksEpochStoreSupplier;
+
+  public NextCandidateForkPostProcessor(
+      CandidateForkConfig nextFork, Supplier<ForksEpochStore> forksEpochStoreSupplier) {
+    this.nextFork = nextFork;
+    this.forksEpochStoreSupplier = forksEpochStoreSupplier;
+  }
+
+  @Override
+  public LedgerAndBFTProof process(
+      LedgerAndBFTProof metadata,
+      EngineStore.EngineStoreInTransaction<LedgerAndBFTProof> engineStore,
+      List<REProcessedTxn> txns)
+      throws PostProcessorException {
+    if (metadata.getProof().getNextValidatorSet().isPresent()
+        && Forks.testCandidate(nextFork, metadata, forksEpochStoreSupplier.get())) {
```

Comment:
Yes, the order of PostProcessors matters. There's only a single PostProcessor in the engine and we use `PostProcessor.combine` to merge two into one.
With the current design we can't really count the votes in either `*ForkPostProcessor` because the latest fork doesn't include any of these post processors, yet when running it we still need to count the votes (with the `CandidateForkVotesPostProcessor`).
I could rename `PostProcessor.combine` to `PostProcessor.append` to make it more explicit and also add a comment where it's actually used. Would that work?

Reply:
I like `PostProcessor.append`, and to add a comment in `CandidateForkVotesPostProcessor` explaining that it has to be run after the other two and why that's guaranteed :)